### PR TITLE
feature master benefeciary

### LINKED ISSUE #1: pallet_supersig Options
# Master Beneficiary Option

Another option for supersig, is the "master beneficiary" option:

## The problem:

The problem is only 3 people in multisig sometimes it's hard to get someone to sign it quickly, but if you had 5 people then its more likely someone will sign your transaction quicker.

## Solution
The funds can go anywhere and be made by anyone, but IF the beneficiary signs it then only one other person needs to sign ELSE IF not beneficiary signs then there has to be a superMajority.

## More Explanation
The most common use for multisig is that there is really one owner and 2 signatories that provide extra security or “approves” transaction. The point of this most common use is that the final destination of the funds is for one Beneficiary. It’s not usually shared funds. 

The limitation with multisig, is that 2 other parties need to be trusted with not stealing the funds. It’s not a major problem with 2/3 multisg, but when we start to go to 2/4, 3/4, 2/5, 3/5, 2/6, 3/6, 4/6, 3/7, 4/7, etc multisigs it increase the risks of funds being taken from the Beneficiary.

Therefore to solve this problem, if the multisig could have a selected Master Beneficiary, and if that beneficary creates a transaction only one other signatory needs to sign. But if that person loses their signature a superMajority can release it.

### PR DESCRIPTION
**Pull request by gdnathan**
close #1

### CODE REVIEW

**Comment by tdelabro** [comment]
Told you in the last PR:
`is_user_in_supersig` should be implemented onto the SuperSig struct. Not taking index as argument.
With this impl if a supersig does not exist this function say that the user is not a member of this supersig. That doesn't exist. Kinda messy. We should first have a supersig, then test if an user is part of it. Not both as the same time, ignoring the None case

You did not do it

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:463]: `nb_removed` vs `number_added`.
Homogenize it

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:590]: rename `remove_members`

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:583]: What if I want to change the role of a member ?
Do we create another extrinsic, or just update it here ?
Second option seems easier for now

You will have to do an `else if current_role != role` and do the insert but not increment number_added

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:607]: You can do this in one single loop:
loop over `members`
use try_mutate
If the value is `NotMember`, don't mutate and go next iteration
if value is not `NotMember`:
- increase `number_removed`
- check `total_members - number_removed >= 1`, if not revert transaction, else remove key from storage

We loop only once.

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/types.rs
Comment (gdnathan) [reply src/types.rs]: I think in rust, enums value are not considered as numbers, so this case doesn't really apply

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=db112c308ca293d10ec3284974f44588)

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs:124]: I'll do this.

Do i manage call nonce and supersig nonce with u32 as well ?

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs:243]: a one member supersig mean we need 1 vote to validate a call.
If the sole member is a master, it mean his vote power is 0 (``1 / 2 = 0`` in rust). So the supersig will be locked, cant be modified, cant event be deleted.

So either a supersig of one user is only for a member and not a master, or a master have a 1 power vote if he is alone. But it is additional checks, so I'd like to ask @decentration if it could be a common use case, to know if it is worth to put those additional checks

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs]: this function add members like it add masters, so i think "add_members" would not be quite correct.
Maybe "add user to supersig" ? I did not find a good name for it

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs:593]: same as add_member function, i think we need a better name. i think `remove_users` is quite good

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs:610]: In a closure (so in a try_mutate), if i return an error, it will not leave the function... so if i have 5 users in the supersig, and want to remove 4, the first three will be removed, and then it will block for the fourth

What i intended to do is that if you want to remove 4 of 5 users, the function will throw an error without removing any users.

Or i can make the function transactional, but if i do, i dont even need to mkae a try_mutate, i can just do the check in my `if ... != Roles::NotMember`

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:519]: Just bellow:

```rust
T::Currency::transfer(
				&supersig_id,
				&beneficiary,
				balance,
				ExistenceRequirement::AllowDeath,
			)?;
```
Maybe use `transfer_all` to be more explicit

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:205]: Rename `CannotDeleteSupersig` so it explain the locked tokens thing.
Better having the name be explicit, than having people read the doc

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:512]: Rename `balance` as `total_balance` because that what it is.

File: src/lib.rs
Comment (tdelabro) [inline src/lib.rs:517]: Here we don't take into account all the money reserved for other calls preimage, that may be waiting for approval.

I suggest we loop over every call stored and unreserve the deposits associated with each preimage before we delete the supersig. Otherwise, the people will just lose their money when we clean up the storage

Comment (tdelabro) [reply src/lib.rs:517]: Make a test that ensure this happen

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs:519]: transfer_all is an extrinsic of pallet_balances, and as i use the trait Currency, and i'm not tightly coupled with pallet_balance, i cannot use it

File: src/lib.rs
Comment (gdnathan) [inline src/lib.rs:708]: `compute_vote_weight` ? `evalutate_vote_weight` ? I think these are better names

File: src/lib.rs
Comment (gdnathan) [inline src/lib.rs:585]: No funds are release when an user leave the supersig

File: src/lib.rs
Comment (gdnathan) [inline src/lib.rs:489]: This should be moved to an internal function, so we can also use it in `leave_supersig`

File: src/lib.rs
Comment (gdnathan) [inline src/lib.rs:111]: Maybe we should rename it `MaxAccountPerTransaction`, to be coherent with the doc comment of the type

File: src/lib.rs
Comment (gdnathan) [inline src/lib.rs:310]: This could fail after a successfull storage mutation. The extrinsic should be marked `#[transactional]`

File: src/lib.rs
Comment (tdelabro) [reply src/lib.rs:687]: En fait il faut eviter les nested transactional: https://github.com/paritytech/substrate/issues/10806
Ce que tu fais c'est que tu mets tous les Extrinsics qui en ont besoin en transactional. Et aucun dans les sous fonctions.

File: src/lib.rs
Comment (tdelabro) [reply src/lib.rs:708]: go for `compute_vote_weight`

File: src/lib.rs
Comment (tdelabro) [reply src/lib.rs:585]: Please do it haha

File: src/lib.rs
Comment (tdelabro) [reply src/lib.rs:489]: Sure !

File: src/lib.rs
Comment (tdelabro) [reply src/lib.rs:310]: Please do

File: src/lib.rs
Comment (tdelabro) [reply src/lib.rs:687]: limite, tous les extrinsics transactionals c'est une bonne regle de base

File: src/lib.rs
Comment (gdnathan) [reply src/lib.rs:111]: i mean this:

```
		/// The maximum number of account that can added or removed in a single call <-- the comment say "account"
		#[pallet::constant]
		type MaxMembersPerTransaction: Get<u32>; // <-- the type say "members"

```

**Pull request merged**
